File: 1670.design-front-middle-back-queue.java
#
# @lc app=leetcode id=1670 lang=java
#
# [1670] Design Front Middle Back Queue
#

import java.util.LinkedList;

# @lc code=start
class FrontMiddleBackQueue {

    LinkedList<Integer> list;

    public FrontMiddleBackQueue() {
        list = new LinkedList<Integer>();
    }

    public void pushFront(int val) {
        list.addFirst(val);
    }

    public void pushMiddle(int val) {
        list.add(list.size() / 2, val);
    }

    public void pushBack(int val) {
        list.addLast(val);
    }

    public int popFront() {
        if (list.isEmpty()) {
            return -1;
        }
        return list.removeFirst();
    }

    public int popMiddle() {
        if (list.isEmpty()) {
            return -1;
        }
        int sz = list.size();
        return list.remove((sz - 1) / 2);
    }

    public int popBack() {
        if (list.isEmpty()) {
            return -1;
        }
        return list.removeLast();
    }
}

/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * FrontMiddleBackQueue obj = new FrontMiddleBackQueue();
 * obj.pushFront(val);
 * obj.pushMiddle(val);
 * obj.pushBack(val);
 * int param_4 = obj.popFront();
 * int param_5 = obj.popMiddle();
 * int param_6 = obj.popBack();
 */
# @lc code=end

File: 1670.design-front-middle-back-queue.go
#
# @lc app=leetcode id=1670 lang=golang
#
# [1670] Design Front Middle Back Queue
#

# @lc code=start
type FrontMiddleBackQueue struct {

	q []int
}


func Constructor() FrontMiddleBackQueue {

	return FrontMiddleBackQueue{q: []int{}}
}


func (this *FrontMiddleBackQueue) PushFront(val int)  {

	this.q = append([]int{val}, this.q...)
}


func (this *FrontMiddleBackQueue) PushMiddle(val int)  {

	pos := len(this.q) / 2

	this.q = append(this.q[:pos], append([]int{val}, this.q[pos:]...)...)
}


func (this *FrontMiddleBackQueue) PushBack(val int)  {

	this.q = append(this.q, val)
}


func (this *FrontMiddleBackQueue) PopFront() int {

	if len(this.q) == 0 {

		return -1

	}

	val := this.q[0]

	this.q = this.q[1:]

	return val
}


func (this *FrontMiddleBackQueue) PopMiddle() int {

	if len(this.q) == 0 {

		return -1

	}

	mid := (len(this.q) - 1) / 2

	val := this.q[mid]

	this.q = append(this.q[:mid], this.q[mid+1:]...)

	return val
}


func (this *FrontMiddleBackQueue) PopBack() int {

	if len(this.q) == 0 {

		return -1

	}

	n := len(this.q)

	val := this.q[n-1]

	this.q = this.q[:n-1]

	return val
}


/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.PushFront(val);
 * obj.PushMiddle(val);
 * obj.PushBack(val);
 * param_4 := obj.PopFront();
 * param_5 := obj.PopMiddle();
 * param_6 := obj.PopBack();
 */
# @lc code=end